src/dataflow/src/render/filter.rs

```diff
@@ -66,134 +322,11 @@ where
-                        }
-
-                        // Convert both of our bounds to `Option<u64>`, where negative numbers
-                        // are advanced up to `Some(0)` and numbers larger than `u64::MAX` are
-                        // set to `None`. These choices are believed correct to narrow intervals
-                        // of `i128` values to potentially half-open `u64` values.
-
-                        // We are "certain" that `lower_bound_i128` is at least `time`, which
-                        // means "non-negative" / not needing to be clamped from below.
-                        let lower_bound_u64 = if lower_bound_i128 > u64::MAX.into() {
-                            None
-                        } else {
-                            Some(u64::try_from(lower_bound_i128).unwrap())
-                        };
-
-                        // We ensured that `upper_bound_i128` is at least `lower_bound_i128`,
-                        // and so it also does not need to be clamped from below.
-                        let upper_bound_u64 = match upper_bound_i128 {
-                            Some(u) if u < 0 => {
-                                panic!("upper bound was ensured at least `time`; should be non-negative");
-                            },
-                            Some(u) if u > u64::MAX.into() => None,
-                            Some(u) => Some(u64::try_from(u).unwrap()),
-                            None => None,
-                        };
-
-                        // Drop to allow the re-use of `data` in the output.
-                        drop(datums_local);
-
-                        // Only proceed if the new time is not greater or equal to upper,
-                        // and if no null values were encountered in bound evaluation.
-                        if lower_bound_u64 != upper_bound_u64 && !null_eval {
-                            // We send `upper` first because it allows us to avoid a clone
-                            // in the case that `upper` is `None`. At this point `lower`
-                            // should not be `None`, and there is nothing lost this order.
-                            let upper_opt = upper_bound_u64.map(|time| Ok((data.clone(), time, -diff)));
-                            let lower_opt = lower_bound_u64.map(|time| Ok((data, time, diff)));
-                            lower_opt.into_iter().chain(upper_opt.into_iter())
-                        } else {
-                            None.into_iter().chain(None.into_iter())
-                        }
-                    } else {
-                        None.into_iter().chain(None.into_iter())
-                    }
+                    let mut datums_local = datums.borrow_with(&data);
+                    // TODO: It would be nice to avoid the `data.clone` for the last output.
+                    let times_diffs = filter_plan.evaluate(&mut datums_local, time, diff);
+                    drop(datums_local);
```

Comment:
maybe add a comment here for why the drop is necessary? (asking because i forgot)

Reply:
Will do (the subsequent closure cannot take ownership of `data` otherwise).